sleap_io/model/instance.py

```diff
@@ -0,0 +1,312 @@
+
+        Raises:
+            ValueError: If a point is associated with a skeleton node
+                name that doesn't exist.
+            TypeError: With a human readable error message, the attribute (of type
+                attrs.Attribute), the expected type, and the value it got.
+        """
+        if points is not None:
+            try:
+                for node_name in points.keys():
+                    if Node(node_name) not in self.skeleton.nodes:
+                        raise KeyError(
+                            f"There is no node named {node_name} in {self.skeleton}"
+                        )
+            except AttributeError:
+                raise TypeError(
+                    "'{name}' must be {type!r} (got {value!r} that is a "
+                    "{actual!r}).".format(
+                        name=attribute.name,
+                        type=dict,
+                        actual=points.__class__,
+                        value=points,
+                    ),
+                    attribute,
+                    dict,
+                    points,
+                )
+
+    @from_predicted.validator
+    def _validate_type_is_PredictedInstance(self, attribute, value):
+        if (value is not None) and (not isinstance(value, PredictedInstance)):
+            raise TypeError(
+                "'{name}' must be {type!r} (got {value!r} that is a "
+                "{actual!r}).".format(
+                    name=attribute.name,
+                    type="Predicted Instance",
+                    actual=value.__class__,
+                    value=value,
+                ),
+                attribute,
+                "Predicted Instance",
+                value,
+            )
+
+    @classmethod
+    def from_pointsarray(
+        cls, points: np.ndarray, skeleton: Skeleton, track: Optional[Track] = None
+    ) -> Instance:
+        """Create an instance from an array of points.
+
+        Args:
+            points: A numpy array of shape `(n_nodes, 2)` and dtype `float32` that
+                contains the points in (x, y) coordinates of each node. Missing nodes
+                should be represented as `NaN`.
+            skeleton: A `sleap.Skeleton` instance with `n_nodes` nodes to associate with
+                the instance.
+            track: Optional `sleap.Track` object to associate with the instance.
+
+        Returns:
+            A new `Instance` object.
+        """
+        predicted_points = dict()
+        node_names: List[str] = [node.name for node in skeleton.nodes]
+        # TODO(LM): Ensure ordering of nodes and points match up.
+        for point, node_name in zip(points, node_names):
+            if (len(point)) == 4:
+                predicted_points[node_name] = Point(
+                    x=point[0],
+                    y=point[1],
+                    visible=bool(point[2]),
+                    complete=bool(point[3]),
+                )
```

Comment:
Is there ever a case where we return less `points` than `node_names`? If so, we may need to add additional check for ordering here and create a test specifically for the ordering of `points`, `node_names`.

Possible case: SLEAP will convert predicted points with a very low confidence score to NaN - are these points saved to the hdf5? If points are stored as a matrix then yes - but if stored as a ragged tensor, then not necessarily and will need to inspect [`Labels.save_file`](https://github.com/talmolab/sleap/blob/8fbaa30b130eec8f2f1e58a19927435d6fbcc4ee/sleap/io/dataset.py#L1946) calls.